### ISSUE TITLE
More Optimization ideas

### ISSUE THREAD

**Issue by torch2424** [torch2424/wasmBoy#73]
Continuation of #33 , made a new issue just to make things easier to follow. That issue has a ton of nice links on what can and can't be done graphically to preserve accuracy :)

~~Should look into replacing cylcle counters, with "cycle targets" In which, we would like something to happen to next for the hardware~~ Would Have to count infinitely :p

Record When the Timer TMA changes in memory and record in a static variable when it does that way we don't have to check if it did every batch process.

Try not to actually implement hardware registers as ISSOtm mentioned in the issue, but rather, calculate them when needed instead.

**Comment by torch2424** [comment]
Try manually profiling with NO performance options enabled and 4X Slowdown on CPU for Link's Awakening Intro:

Runs normally at around 28-30fps.

* commenting `_drawScanline()` ~ + 10 fps
* commenting `updateSound()` ~ + 10fps
* commenting `updateTimers()` ~ + 0 fps
* commenting `checkInterrupts()` + ~ 4 fps (Game doesn't run though duh)
*  commenting `let readTrapResult: i32 = checkReadTraps(offset);`
* skipping `checkWriteTraps()` ~ Doesn't work
* commenting  both `_drawScanline()` and `updateSound()` ~ + 15fps

**Comment by torch2424** [comment]
Ask Dcode if it would be worth storing variables as static class rather than in memory, and essentially that is what happens anyways, so there would be no benefit 😢

**Comment by torch2424** [comment]
Look into profiling on nightly firefox to hopefully get some better profiling? https://github.com/WebAssembly/binaryen/issues/1070

**Comment by torch2424** [comment]
Try running asm.js output again for better profiling?

**Comment by torch2424** [comment]
Try to convert more variables/functions to i32

**Comment by torch2424** [comment]
<img width="906" alt="screen shot 2018-04-18 at 11 05 17 am" src="https://user-images.githubusercontent.com/1448289/38949828-6a96e1e8-42f8-11e8-96d5-bdc0495d6da2.png">

**Comment by torch2424** [comment]
Try to cache branch results where possible

**Comment by torch2424** [comment]
Cache hardware register properties.

For example, create static classes for hardware registers. And when they are written to, simply update the static class, but allow the write. Thjat way the CPU is happy, and we are happy because we aren't constantly checking single bits on the same thing, and loading the same thing from memory over and over. 😄

**Comment by torch2424** [comment]
See above, try to cach echeckBitOnByte results elsewhere

**Comment by torch2424** [comment]
Refactor CPU code to used shared functions, and the refactor those functions :)

**Comment by torch2424** [comment]
Some Assemblyscript optimizations mentioned by @CryZe

<img width="605" alt="screen shot 2018-04-19 at 10 35 12 am" src="https://user-images.githubusercontent.com/1448289/39008430-7f193a88-43bd-11e8-836b-310ba9fd0869.png">

<img width="1030" alt="screen shot 2018-04-19 at 10 35 22 am" src="https://user-images.githubusercontent.com/1448289/39008436-830fa686-43bd-11e8-8f59-f8a8c69fa2b0.png">

**Comment by torch2424** [comment]
Inlining functions when `-03`does not?

**Comment by torch2424** [comment]
Okay okay okay okay okay...

_**Story time**_

I had an idea, and was like, huh, Maybe I should try some of the JS emus profile them, and see how that goes.

So I tried [Gameboy Online](https://github.com/taisel/GameBoy-Online), and like usual, would run 60fps on my 2015 MBP with 4x CPU slowdown on chrome, Which @binji was nice enough to confirm would affect wasm performance. Also, once they and I were talking, and found out [browsers used to tune their JS engines to gameboy emulators like Gameboy Online](https://github.com/chromium/octane/blob/master/gbemu-part1.js), so it makes sense why it would run so great.

But after profiling that, had a hard time reading the source since it follows the monolithic core pattern, and looked elsewhere. So first I tried one of my favorite reference emus, [node-gameboy](https://www.npmjs.com/package/node-gameboy). I did the 4x slowdown, and guess what? It was probably running at only about 20fps (no fps counter). So I got suspicious, and I found this [list of other js emus](https://github.com/fcambus/jsemu), and started going through them, such as [jsGbc](https://ardean.github.io/jsGBC-web/) and [jsNES](https://jsnes.org/). If you 4x slowdown them, guess what? You will probably get about 20-45 fps it seems (no fps counter).

Okay, so then I thought, "Well awesome! My emu isn't worse than other similar options, it is nice wasm actually gives you that nice perf bump!" But, I thought, "I wonder how I would compare to other wasm emulators, that are not binjigb". So I tried [PlutoBoy](https://github.com/RossMeikleham/PlutoBoy), and [SaltyNES](https://github.com/workhorsy/SaltyNES), and [Pinky](https://github.com/koute/pinky) and they all run about the same as mine (if not a little less) at 4x slowdown on my same machine.

So the only exceptions I can find are [binjigb](https://github.com/binji/binjgb) and the super recently uploaded but very incomplete which is why it is fast [wasm-gb](https://github.com/andrewimm/wasm-gb) that run well even at 6x slowdown. One thing I noticed is that they both seem to leverage llvm. so that may be the culprit behind Super fast Wasm in the browser. Where [Gameboy Online](https://github.com/taisel/GameBoy-Online) has source code that uses some JIT compiling, and was fine tuned by browsers to run hella fast.

**TL;DR** I am not a "completely" horrible programmer, who writes super inefficient code. The rate at which the emulator currently runs is a little better than average, and should be expected 😄
